src/planner/expand_hypertable.c

```diff
@@ -327,10 +337,53 @@ constify_timestamptz_op_interval(PlannerInfo *root, OpExpr *constraint)
 									constraint->inputcollid);
 }
 
+static Datum
+calculate_timestamptz_timezone_bucket(int strategy, Datum width, Datum value, List *args)
+{
+	Assert(list_length(args) == 5);
+
+	if (TIMESTAMP_NOT_FINITE(DatumGetTimestampTz(value)))
+		return value;
+
+	Const *tzname = castNode(Const, lthird(args));
+	Const *offset = castNode(Const, lfifth(args));
+	Datum new_timestamp;
+	bool have_offset = (!offset->constisnull);
+
+	new_timestamp = DirectFunctionCall2(timestamptz_zone, tzname->constvalue, value);
+
+	if (have_offset)
+	{
+		/* Apply offset. */
+		new_timestamp =
+			DirectFunctionCall2(timestamp_mi_interval, new_timestamp, offset->constvalue);
+	}
+
+	/*
+	 * If the interval has a day component then the calculation needs
+	 * to take into account daylight saving time switches and thereby a
+	 * day would not always be exactly 24 hours. We mitigate this by
+	 * adding a safety buffer to account for these dst switches when
+	 * dealing with intervals with day component. These calculations
+	 * will be repeated with exact values during execution.
+	 * Since dst switches seem to range between -1 and 2 hours we set
+	 * the safety buffer to 4 hours.
+	 */
+	TimestampTz new_value = DatumGetTimestampTz(new_timestamp);
+
+	if (strategy == BTLessStrategyNumber || strategy == BTLessEqualStrategyNumber)
+		new_value += 4 * USECS_PER_HOUR;
+	else if (strategy == BTGreaterStrategyNumber || strategy == BTGreaterEqualStrategyNumber)
+		new_value -= 4 * USECS_PER_HOUR;
+
+	return TimestampTzGetDatum(new_value);
```

Comment:
I think return new_value; should do the job.

Reply:
We're returning a Datum here, this is why I'm doing it.

Reply:
Ok.